Loading Data set

In [89]:
from sklearn.datasets import load_iris
from sklearn.linear_model import LogisticRegression
from sklearn.model_selection import train_test_split

iris = load_iris(as_frame=True)

In [90]:
iris.target_names

array(['setosa', 'versicolor', 'virginica'], dtype='<U10')

In [91]:
iris.data

sepal length (cm)  sepal width (cm)  petal length (cm)  petal width (cm)
0                  5.1               3.5                1.4               0.2
1                  4.9               3.0                1.4               0.2
2                  4.7               3.2                1.3               0.2
3                  4.6               3.1                1.5               0.2
4                  5.0               3.6                1.4               0.2
..                 ...               ...                ...               ...
145                6.7               3.0                5.2               2.3
146                6.3               2.5                5.0               1.9
147                6.5               3.0                5.2               2.0
148                6.2               3.4                5.4               2.3
149                5.9               3.0                5.1               1.8

[150 rows x 4 columns]

In [92]:
iris.target_names
print(iris.DESCR)

.. _iris_dataset:

Iris plants dataset
--------------------

**Data Set Characteristics:**

    :Number of Instances: 150 (50 in each of three classes)
    :Number of Attributes: 4 numeric, predictive attributes and the class
    :Attribute Information:
        - sepal length in cm
        - sepal width in cm
        - petal length in cm
        - petal width in cm
        - class:
                - Iris-Setosa
                - Iris-Versicolour
                - Iris-Virginica
                
    :Summary Statistics:

    ============== ==== ==== ======= ===== ====================
                    Min  Max   Mean    SD   Class Correlation
    ============== ==== ==== ======= ===== ====================
    sepal length:   4.3  7.9   5.84   0.83    0.7826
    sepal width:    2.0  4.4   3.05   0.43   -0.4194
    petal length:   1.0  6.9   3.76   1.76    0.9490  (high!)
    petal width:    0.1  2.5   1.20   0.76    0.9565  (high!)
    ============== ==== ==== ======= ===== ===========

In [93]:
X = iris.data
y = iris.target_names[iris.target] == 'virginica'
X_train, X_test, y_train, y_test = train_test_split(X, y,train_size=0.75, random_state=42)

In [94]:
LR = LogisticRegression(random_state=42)
LR.fit(X_train, y_train)

LogisticRegression(random_state=42)

In [95]:
y_test

array([False, False,  True, False, False, False, False,  True, False,
       False,  True, False, False, False, False, False,  True, False,
       False,  True, False,  True, False,  True,  True,  True,  True,
        True, False, False, False, False, False, False, False,  True,
       False, False])

In [96]:
y_predict_test = LR.predict(X_test)
y_predict_test

array([False, False,  True, False, False, False, False,  True, False,
       False,  True, False, False, False, False, False,  True, False,
       False,  True, False,  True, False,  True,  True,  True,  True,
        True, False, False, False, False, False, False, False,  True,
       False, False])

In order to evaluate logistic regression we use confusion matrix.

In [97]:
from sklearn.metrics import confusion_matrix
cnf_matrix = confusion_matrix(y_test, y_predict_test)
print(cnf_matrix)

[[26  0]
 [ 0 12]]


There are no false positive and false negatives.

Cross validation

In [102]:
from sklearn.model_selection import cross_val_score
scores = cross_val_score(LR, X_train, y_train, cv=10, scoring='accuracy')
print(scores)
scores.mean()

[1.         1.         0.90909091 1.         0.81818182 0.90909091
 1.         1.         1.         0.90909091]


0.9545454545454545

Precision and recall

In [99]:
from sklearn.metrics import precision_score
prec_score = precision_score(y_test, y_predict_test)
prec_score

1.0

In [100]:
from sklearn.metrics import recall_score
recall = recall_score(y_test, y_predict_test)
recall

1.0

The model is giving 1 for precision and recall. And in cross validation using 10 split it gives an accuarcy mean of 95.45%. 